umbraco/Umbraco-CMS
# V13: Add Guids to Users

Adds an actual GUID as a key to users instead of the faked one. 

One of the problems implementing this, is that by adding a new column, you were unable to log in to authorize the upgrade, since we'd error out because we expected the Key column to be there. 

To get around this I've added a new `GetForUpgradeBy*` this will allow you to request a user by id, email or username, as this is needed by identity, these new methods will only return as user while in upgrade, the goal is to have a method that fetches the bare minimum to authorize the upgrade. 

Another issue is that the user already had a fake key, that was used to resolve external logins and 2fa providers, to facility this transition to proper keys I've made a `ResolveEntityIdFromIdentityId` which will resolve the umbraco Entity `IUser` or `IMember` fro the ID in the `IdentityUser`, this ID can depending the situation be: 

1. An int id - this means the user logged in "normally"
2. A proper GUID - this means the user is logging in using an external login provider, and has migrated
3. A fake GUID - This means the user is authorizing the migration and logged in using an external login provider 

The method handles all of these cases. 

I've also added an internal `IsFakeGuid` which will determine if a GUID is a true GUID, or our fake ones. 


Lastly I had mistakenly done `Guid.NewGuid` when creating the user groups, and I notices some that did `1.ToGuid()`, when creating data, I've changed these to match the same pattern as the rest.


## Testing

In testing I tested that I could 

1. Set up External Login before this branch
2. Change to this branch
3. Ensure that I can log in both normally and with external login when authorizing the upgrade
4. Ensure that I can log in with 2fa enabled to authorize the upgrade. 
5. Ensure that I can log in after the upgrade, both with external login and using 2fa¨
